### PR TITLE
Support for search_as_you_type field type and bool_prefix query type

### PR DESCRIPTION
**Pull request by rickardp**
The search_as_you_type field type was introduced in ElasticSearch 7.2. We missed this functionality in NEST so we went ahead and implemented it.

Also added the bool_prefix query type that we needed to implement our search completions.

Reported in issue #4033

### CODE REVIEW

File: src/Examples/Examples/Mapping/Types/SearchAsYouTypePage.cs [hunk 1]
```diff
@@ -1,18 +1,23 @@
...
 	{
-		[U(Skip = "Example not implemented")]
+		[U]
 		public void Line18()
 		{
 			// tag::6f31f9cfe0dd741ccad4af62ba8f815e[]
-			var response0 = new SearchResponse<object>();
+			var createIndexResponse = client.Indices.Create("my_index", c => c
+				.Map(m => m
+					.Properties(p => p.SearchAsYouType(t => t.Name("my_field")))
```
Comment (russcam) [inline src/Examples/Examples/Mapping/Types/SearchAsYouTypePage.cs]: Could you put the `.SearchAsYouType(t => t` and `.Name("my_field"))` on separate, indented lines? This example will go into the Elasticsearch reference documentation, so it'll help for consistency

File: src/Examples/Examples/Mapping/Types/SearchAsYouTypePage.cs [hunk 2]
```diff
@@ -37,14 +42,20 @@ public void Line72()
...
 		{
 			// tag::9bd25962f177e86dbc5a8030a420cc31[]
-			var response0 = new SearchResponse<object>();
+			var query = client.Search<object>(s => s
+				.Index("my_index")
+				.Query(q => q
+					.MultiMatch(mm => mm
+						.Query("brown f")
+						.Type(TextQueryType.BoolPrefix)
+						.Fields(f => f.Field("my_field").Field("my_field._2gram").Field("my_field._3gram")))));
```
Comment (russcam) [inline src/Examples/Examples/Mapping/Types/SearchAsYouTypePage.cs]: Could the individual `.Field(...)` calls be put on separate lines?

File: src/Examples/Examples/Mapping/Types/SearchAsYouTypePage.cs [hunk 3]
```diff
@@ -60,14 +71,19 @@ public void Line89()
 			}");
 		}
 
-		[U(Skip = "Example not implemented")]
+		[U(Skip = "Is doc correct here?")]
 		public void Line151()
 		{
 			// tag::0ced86822f8c0a479af5e1fe28dfc2ec[]
-			var response0 = new SearchResponse<object>();
+			var query = client.Search<Dictionary<string, string>>(s => s
```
Comment (russcam) [inline src/Examples/Examples/Mapping/Types/SearchAsYouTypePage.cs]: I think just `object` is better, with `searchResponse` as variable name

File: src/Examples/Examples/Mapping/Types/SearchAsYouTypePage.cs [hunk 4]
```diff
@@ -60,14 +71,19 @@ public void Line89()
...
+				.Index("my_index")
+				.Query(q => q
+					.MatchPhrasePrefix(pp => pp
+						.Field(doc => doc["my_field"])
```
Comment (russcam) [inline src/Examples/Examples/Mapping/Types/SearchAsYouTypePage.cs]:
```suggestion
						.Field("my_field")
```

File: src/Examples/Examples/Mapping/Types/TextPage.cs [hunk 5]
```diff
@@ -5,14 +5,18 @@ namespace Examples.Mapping.Types
...
 	{
-		[U(Skip = "Example not implemented")]
+		[U]
 		public void Line22()
 		{
 			// tag::24ea1c6cdf10165228951e562b7ec0ef[]
-			var response0 = new SearchResponse<object>();
+			var createIndexResponse = client.Indices.Create("my_index", c => c
+				.Map(m => m
+					.Properties(p => p.Text(t => t.Name("full_name")))
```
Comment (russcam) [inline src/Examples/Examples/Mapping/Types/TextPage.cs]: Put `.Text(...)` and `.Name(...)` on separate lines, for style consistency

File: src/Nest/Mapping/Types/Core/Text/TextProperty.cs [hunk 6]
```diff
@@ -3,59 +3,58 @@
 using System.Runtime.Serialization;
 using Elasticsearch.Net.Utf8Json;
 
-namespace Nest
-{
+namespace Nest {
```
Comment (russcam) [inline src/Nest/Mapping/Types/Core/Text/TextProperty.cs]: The editorconfig file should format block braces to separate lines with `CTRL + E` `CTRL + F`

**Comment by russcam** [comment]
I'm looking to pull this in now @rickardp, so I'm gonna make the changes requested. Hope this OK

**Comment by russcam** [comment]
Digging into this a bit further. it looks like all properties of [`TextProperty`](https://www.elastic.co/guide/en/elasticsearch/reference/7.2/text.html#text-params) are not supported by [`SearchAsYouTypeProperty`](https://www.elastic.co/guide/en/elasticsearch/reference/7.2/search-as-you-type.html#general-params) e.g. `eager_global_ordinals`, `fielddata`, etc. Because of this, I don't think it can inherit from `TextProperty`

**Comment by rickardp** [comment]
It should be easy to change this, keeping only the common properties in the shared class. I think a new interface is needed to define what is shared. I can have a look at it later today.

File: src/Nest/Mapping/Types/Core/Text/TextProperty.cs (hunk 6 above)
Comment (rickardp) [reply src/Nest/Mapping/Types/Core/Text/TextProperty.cs]: Yes, I saw this. I think VS 2019 is not happy about some format in the .editorconfig, which causes formatting to actually make things worse. I’ve now fixed the .editorconfig locally. If there is interest I can submit a new PR.

**Comment by russcam** [comment]
I started on making the changes locally. I should be able to push something up for them

**Comment by russcam** [comment]
Have to be careful in refactoring types that it doesn't result in a _binary breaking change_; the client maintains backwards binary compatibility within a major version.

File: src/Nest/Mapping/Types/Core/Text/TextProperty.cs (hunk 6 above)
Comment (russcam) [reply src/Nest/Mapping/Types/Core/Text/TextProperty.cs]: Please feel free to open a PR for the editorconfig 👍

**Comment by russcam** [comment]
@rickardp I've opened https://github.com/elastic/elasticsearch-net/pull/4048

**Comment by russcam** [comment]
I'm going to close this as #4048 has now been merged; I flagged you as Co-Author.

Thanks again for opening this PR, @rickardp 👍

**Pull request closed without merging**